### PR TITLE
feat: deprecate endpoint option in favor of botProviderEndpoint

### LINKED ISSUE #66: Deprecate `endpoint` option in favor of `botProviderEndpoint`
## Problem Statement

The current SDK configuration requires two separate endpoint configurations:
- `endpoint`: The full SSE endpoint URL 
- `botProviderEndpoint`: The base bot provider URL

This creates redundancy since `endpoint` is always `botProviderEndpoint + '/message/sse'`. Having two configurations for what is essentially the same service increases complexity and potential for errors.

## Proposed Solution

Deprecate the `endpoint` option and derive it automatically from `botProviderEndpoint`. This will:
1. Simplify configuration
2. Reduce potential for errors
3. Provide a single source of truth
4. Maintain backward compatibility during transition

## Example

**Current (redundant):**
```typescript
config: {
  endpoint: 'https://api.example.com/ns/xxx/bot-provider/bp-xxx/message/sse',
  botProviderEndpoint: 'https://api.example.com/ns/xxx/bot-provider/bp-xxx'
}
```

**Proposed (simplified):**
```typescript
config: {
  botProviderEndpoint: 'https://api.example.com/ns/xxx/bot-provider/bp-xxx'
  // endpoint automatically derived as botProviderEndpoint + '/message/sse'
}
```

## Detailed Proposal

A comprehensive deprecation proposal has been added to the repository:
[View full proposal](https://github.com/asgard-ai-platform/asgard-js-sdk/blob/main/docs/deprecation-endpoint-to-botproviderendpoint.md)

The proposal includes:
- Detailed implementation plan
- Code examples
- Migration guide
- Timeline
- Backward compatibility strategy

## Request for Feedback

Please review the proposal and provide feedback on:
1. The deprecation timeline
2. Any edge cases not considered
3. Impact on existing integrations
4. Alternative approaches

## Labels
- enhancement
- breaking-change (for future major version)
- discussion

### PR DESCRIPTION
**Pull request by purepennons**
## Summary

This PR implements the deprecation of the `endpoint` configuration option in favor of using `botProviderEndpoint` as the single source of truth for bot provider service URLs, as outlined in issue #66. Additionally, it includes robustness improvements to prevent app crashes when configuration is incomplete and enhanced TypeScript type safety.

### Changes Made

- **Core Package Updates (`@asgard-js/core`)**:
  - Updated `ClientConfig` to use TypeScript discriminated union for compile-time type safety
  - Modified `AsgardServiceClient` constructor to automatically derive SSE endpoint from `botProviderEndpoint`
  - Added deprecation warning in debug mode when `endpoint` is used
  - Improved endpoint derivation with proper trailing slash handling
  - Maintained both compile-time and runtime validation for defense-in-depth
  - Maintained backward compatibility - both configuration options work during transition

- **React Package Updates (`@asgard-js/react`)**:
  - Enhanced `getBotProviderModels` to handle missing `botProviderEndpoint` gracefully
  - Added console warning when `botProviderEndpoint` is not provided
  - Return valid stub functions with proper data structures to prevent crashes
  - Ensure theme system compatibility with empty annotation structures

- **Documentation**:
  - Enhanced core package README with comprehensive API documentation
  - Added detailed deprecation proposal document explaining migration path
  - Updated README to clarify behavior when both endpoints are provided

- **Testing**:
  - Set up comprehensive test environment for both core and react packages
  - Added tests for trailing slash handling and deprecation warnings
  - Updated tests to verify both TypeScript compile-time and runtime validation
  - Ensured backward compatibility with existing configurations

### Breaking Changes

- **None in this release** - This is a backward-compatible change
- `endpoint` is now deprecated but continues to work with deprecation warnings
- Future major version will remove `endpoint` entirely

### Migration Path

**Before:**
```typescript
const client = new AsgardServiceClient({
  endpoint: 'https://api.example.com/ns/xxx/bot-provider/bp-xxx/message/sse',
  botProviderEndpoint: 'https://api.example.com/ns/xxx/bot-provider/bp-xxx',
  apiKey: 'xxx'
});
```

**After (Recommended):**
```typescript
const client = new AsgardServiceClient({
  botProviderEndpoint: 'https://api.example.com/ns/xxx/bot-provider/bp-xxx',
  apiKey: 'xxx'
  // endpoint is automatically derived as botProviderEndpoint + '/message/sse'
});
```

### TypeScript Type Safety Improvements

**Enhanced Compile-time Validation:**
- Implemented discriminated union type that enforces at compile-time that either `endpoint` or `botProviderEndpoint` must be provided
- TypeScript will now catch invalid configurations during development
- Improved developer experience with better IDE autocomplete and error messages

**Defense-in-Depth Approach:**
- Maintained runtime validation alongside TypeScript compile-time checks
- Provides safety for JavaScript usage and dynamic configurations
- Clear error messages for debugging

### Robustness Improvements

**Graceful Handling of Missing botProviderEndpoint:**
- Apps using only the deprecated `endpoint` prop will no longer crash
- Console warning alerts developers about missing bot provider features
- Theme system continues to work with empty/default values
- Backward compatibility maintained for existing integrations

### Transition Period Behavior

During the transition period, all approaches work seamlessly:
- **Only `botProviderEndpoint` provided**: SSE endpoint automatically derived (recommended)
- **Only `endpoint` provided**: Works with deprecation warning when `debugMode` is enabled
- **Both provided**: `endpoint` takes precedence with deprecation warning
- **Neither provided**: TypeScript compilation error + runtime error for comprehensive safety

### Benefits

1. **Single Source of Truth**: Only one endpoint configuration needed
2. **Enhanced Type Safety**: Compile-time validation prevents configuration errors
3. **Reduced Configuration Complexity**: Fewer chances for configuration errors
4. **Robust Error Handling**: Apps continue to work even with incomplete configuration
5. **Backward Compatibility**: Existing code continues to work during transition
6. **Clear Migration Path**: Deprecation warnings guide users to new approach
7. **Developer Experience**: Better IDE support and error messages
8. **Defense in Depth**: Both compile-time and runtime validation

### Technical Improvements

- **TypeScript Discriminated Union**: Enforces configuration constraints at compile time
- **String Interpolation Fix**: Corrected deprecation warning message formatting
- **Trailing Slash Handling**: Proper URL concatenation prevents double slashes
- **Type Safety**: All stub functions return proper TypeScript interfaces
- **Theme Compatibility**: Empty annotation structures work with existing theme merging
- **Runtime Safety**: Maintained runtime validation for comprehensive error handling

Closes #66

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,12 +6,22 @@
...
     "watch:core": "nx build @asgard-js/core --watch",
     "lint:react": "nx run @asgard-js/react:lint",
     "test:react": "nx run @asgard-js/react:test",
+    "test:react:watch": "cd packages/react && npm run test:watch",
+    "test:react:ui": "cd packages/react && npm run test:ui",
+    "test:react:coverage": "cd packages/react && npm run test:coverage",
     "build:react": "nx run @asgard-js/react:build --skip-nx-cache --verbose",
     "watch:react": "nx build @asgard-js/react --watch",
+    "test": "nx run-many --target=test --projects=@asgard-js/core,@asgard-js/react",
+    "test:packages": "nx run-many --target=test --projects=@asgard-js/core,@asgard-js/react",
```
[the remark this answers is not in the document]
Comment (purepennons) [reply package.json]: resolved

File: packages/core/src/lib/client.ts [hunk 2]
```diff
@@ -13,21 +13,37 @@ import { EventEmitter } from './event-emitter';
...
+      // Use provided endpoint but warn about deprecation
+      this.endpoint = config.endpoint;
+      if (this.debugMode) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          '[AsgardServiceClient] The "endpoint" option is deprecated and will be removed in the next major version. ' +
+          'Please use "botProviderEndpoint" instead. The SSE endpoint will be automatically derived as "${botProviderEndpoint}/message/sse".'
+        );
+      }
+    }
```
Comment (purepennons) [inline packages/core/src/lib/client.ts:46]: The current implementation has a potential issue with the deprecation warning message. The template literal syntax `${botProviderEndpoint}` inside the string won't be interpolated. Consider updating this to properly show the derived endpoint:

```typescript
console.warn(
  '[AsgardServiceClient] The "endpoint" option is deprecated and will be removed in the next major version. ' +
  `Please use "botProviderEndpoint" instead. The SSE endpoint will be automatically derived as "${config.botProviderEndpoint}/message/sse".`
);
```

*Generated by Claude Code*

File: packages/core/src/lib/client.spec.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+      
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      expect((client as any).apiKey).toBe('test-key');
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      expect((client as any).debugMode).toBe(true);
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      expect((client as any).transformSsePayload).toBe(transformSsePayload);
+    });
+  });
+});
```
Comment (purepennons) [inline packages/core/src/lib/client.spec.ts:124]: Good test coverage! However, consider adding a test case that verifies the deprecation warning message actually contains the correct interpolated endpoint value when `debugMode` is true. This would ensure the string interpolation issue mentioned above is caught by tests.

Example:
```typescript
it('should show deprecation warning with correct endpoint when using deprecated endpoint option', () => {
  const config: ClientConfig = {
    endpoint: 'https://api.example.com/custom/sse',
    botProviderEndpoint: 'https://api.example.com/bot-provider/bp-123',
    apiKey: 'test-key',
    debugMode: true,
  };

  new AsgardServiceClient(config);
  
  expect(consoleSpy).toHaveBeenCalledWith(
    expect.stringContaining('https://api.example.com/bot-provider/bp-123/message/sse')
  );
});
```

*Generated by Claude Code*

File: packages/core/src/lib/client.ts [hunk 4]
```diff
@@ -13,21 +13,37 @@ import { EventEmitter } from './event-emitter';
...
+    // Handle endpoint derivation and deprecation
+    if (!config.endpoint && config.botProviderEndpoint) {
+      // Derive endpoint from botProviderEndpoint (new recommended way)
+      this.endpoint = `${config.botProviderEndpoint}/message/sse`;
+    } else if (config.endpoint) {
+      // Use provided endpoint but warn about deprecation
+      this.endpoint = config.endpoint;
+      if (this.debugMode) {
+        // eslint-disable-next-line no-console
+        console.warn(
```
Comment (purepennons) [inline packages/core/src/lib/client.ts:41]: Consider adding a type guard or validation to ensure `botProviderEndpoint` doesn't end with a trailing slash before appending `/message/sse`. This would prevent potential double slashes in the derived endpoint:

```typescript
// Derive endpoint from botProviderEndpoint (new recommended way)
const baseEndpoint = config.botProviderEndpoint.endsWith('/') 
  ? config.botProviderEndpoint.slice(0, -1) 
  : config.botProviderEndpoint;
this.endpoint = `${baseEndpoint}/message/sse`;
```

*Generated by Claude Code*

File: README.md [hunk 5]
```diff
@@ -92,13 +88,42 @@ const App = () => {
...
+  botProviderEndpoint: 'https://api.asgard-ai.com/ns/{namespace}/bot-provider/{botProviderId}',
+  // SSE endpoint is automatically derived as: botProviderEndpoint + '/message/sse'
+}
+```
+
+**Benefits:**
+- Simplified configuration with single endpoint
+- Reduced chance of configuration errors
+- Automatic endpoint derivation
+
```
Comment (purepennons) [inline README.md:117]: The migration guide is well-written! Consider adding a note about what happens if both `endpoint` and `botProviderEndpoint` are provided during the transition period. Based on the implementation, it seems `endpoint` takes precedence, which users should be aware of:

```markdown
#### Transition Period
During the transition period, both approaches will work:
- If only `botProviderEndpoint` is provided: The SSE endpoint will be automatically derived
- If only `endpoint` is provided: It will work but show a deprecation warning
- If both are provided: The `endpoint` will be used with a deprecation warning (endpoint takes precedence)
```

*Generated by Claude Code*

**Review by purepennons — COMMENTED** [review]
## Review Summary

This PR successfully implements the deprecation of the `endpoint` option in favor of `botProviderEndpoint`. The implementation is backward-compatible and includes comprehensive documentation updates. However, there are a few issues that should be addressed:

### Key Issues:

1. **String interpolation bug**: The deprecation warning message uses template literal syntax inside a regular string, which won't interpolate the `botProviderEndpoint` value correctly.

2. **Missing edge case handling**: The implementation doesn't handle trailing slashes in `botProviderEndpoint`, which could result in double slashes in the derived endpoint.

3. **Test coverage gap**: While the tests are comprehensive, they don't verify that the deprecation warning message contains the correct interpolated values.

4. **Documentation clarity**: The README should clarify what happens when both `endpoint` and `botProviderEndpoint` are provided.

### Positive Aspects:
- Excellent backward compatibility approach
- Clear migration path with helpful warnings
- Comprehensive documentation updates
- Well-structured test suite

Please address the string interpolation issue and consider the other suggestions to improve the robustness of this deprecation implementation.

*Generated by Claude Code*

**Comment by purepennons** [comment]
All review comments have been addressed in commit 94be458:

✅ **Fixed string interpolation bug** - The deprecation warning now uses a generic message without template literal syntax issues
✅ **Added trailing slash handling** - Using `replace(/\/+$/, '')` to remove all trailing slashes and prevent double slashes
✅ **Added comprehensive test coverage** - New tests verify trailing slash handling and proper deprecation warnings
✅ **Updated README documentation** - Added clarification about precedence when both endpoints are provided

The implementation now properly handles all the edge cases identified in the review. All tests are passing.

*Generated by Claude Code*

File: packages/react/src/models/bot-provider.ts [hunk 6]
```diff
@@ -28,7 +28,46 @@ export const getBotProviderModels = (
...
+        managedFields: [],
+      };
+    };
+
+    const stubGetAnnotations = async (): Promise<Record<string, unknown>> => {
+      const metadata = await stubGetAsgardBotProviderMetadata();
+
+      return annotationSelectorFromBotProviderMetadata(metadata);
+    };
+
```
[the remark this answers is not in the document]
Comment (purepennons) [reply packages/react/src/models/bot-provider.ts]: fixed

File: packages/core/src/types/client.ts [hunk 7]
```diff
@@ -33,7 +33,15 @@ export interface SseHandlers {
...
+  /**
+   * @deprecated Use `botProviderEndpoint` instead. This will be removed in the next major version.
+   * If provided, it will be used. Otherwise, it will be automatically derived as `${botProviderEndpoint}/message/sse`
+   */
+  endpoint?: string;
+  /**
+   * Base URL for the bot provider service.
+   * The SSE endpoint will be automatically derived as `${botProviderEndpoint}/message/sse`
+   */
   botProviderEndpoint?: string;
```
[the remark this answers is not in the document]
Comment (purepennons) [reply packages/core/src/types/client.ts]: fixed
